### ISSUE TITLE
Proper Row object rather than Vec<Value>?

### ISSUE THREAD

**Issue by kardeiz** [blackbeam/rust-mysql-simple#38]
It would be nice if the `QueryResult` iterator yielded a proper `Row` object rather than a simple `Vec<Value>`. That is, it would be convenient if each row contained information about the query columns/statement. My specific use case would be a `get` method on `Row` that could take a `&str` (column name) or `usize` parameter, and return the `Value` for that position.

Both [rusqlite](https://github.com/jgallagher/rusqlite/blob/master/src/lib.rs#L1107) and the [postgres driver](https://github.com/sfackler/rust-postgres/blob/master/src/rows.rs#L188) provide such an interface. I haven't put together a pull request since this might be a radical change to the API, and I don't know if there are performance/overhead concerns...

Just for fun, here is a workaround I am currently using (though it is kind of ugly and probably inefficient):

```rust
pub struct MyWrapper<'a>(QueryResult<'a>);

pub struct MyRow {
  pub row: RefCell<HashMap<usize, Value>>,
  pub columns: HashMap<String, usize>
}

impl MyRow {

  pub fn new(row: MyResult<Vec<Value>>, columns: &[Column]) -> Self {
    let _row = row.into_iter()
      .flat_map(|x| x)
      .enumerate()
      .collect();
    let _columns = columns.iter()
      .map(|x| x.name.clone() )
      .filter_map(|x| String::from_utf8(x).ok() )
      .enumerate()
      .map(|(x, y)| (y, x) )
      .collect();
    MyRow { row: RefCell::new(_row), columns: _columns }
  }

  fn at<T: FromValue>(&self, s: &str) -> Option<T> {
    self.columns.get(s).into_iter()
      .filter_map(|i| self.row.borrow_mut().remove(&i) )
      .filter_map(|x| from_value_opt(x).ok() )
      .next()
  }

}

impl<'a> Iterator for MyWrapper<'a> {
  type Item = MyRow;

  fn next(&mut self) -> Option<MyRow> {
    self.0.next()
      .map(|row| MyRow::new(row, self.0.columns_ref() ) )
  }
}

pub trait MyWrapped<'a> {
  fn wrapped(self) -> MyWrapper<'a>;
}

impl<'a> MyWrapped<'a> for QueryResult<'a> {
  fn wrapped(self) -> MyWrapper<'a> { MyWrapper(self) }
}
```

Which I can use like:

```rust
impl User {
  pub fn all(pool: &MyPool) -> Vec<User> {
    pool.prep_exec("SELECT * FROM users", ()).into_iter()
      .flat_map(|x| x.wrapped() )
      .map(|row| User { email: row.at("email"), ..Default::default() } )
      .collect()
  }
}
```

**Comment by blackbeam** [comment]
Yeah, i think this workaround would be quite inefficient.
Also looks like it's efficiency bound to a number of rows because it will call HashMap::get on each row [which is slow](https://stackoverflow.com/questions/32071415/why-is-this-hashmap-search-slower-than-expected) AFAIK.

I usually prefer something like this:

```rust
let users: Vec<User> = conn.prep_exec("SELECT * FROM users", ()).map(|result| {
    let email_idx = result.column_index("email").unwrap();
    result.into_iter().flat_map(|row| row).map(|mut row| {
        User {
            email: from_value(row.remove(email_idx)),
            ..User::default()
    }).collect()
});
```

But anyway i think you are right and i should wrap Vec<Value> to a something with better API.
I have 2.0 branch in my local repo with a bunch of additions, but have problems with spare time last few months.

**Comment by kardeiz** [comment]
Thanks for the comment. I'm just now coming back to Rust and your library--

I really like what you've done in 2.1.0+. The `Row` object is nice and the `column_indexes` method on `QueryResult` is very helpful for my use case.

I still think it would be nice to have some reference to the columns available to the `Row`. Since we can't return a reference to `self.columns` in the `QueryResult` iterator, what would be the implications of doing something like [this](https://github.com/kardeiz/rust-mysql-simple/compare/master...kardeiz:gh-issue-38?expand=1)? 

(I have no idea how `Arc` affects performance, etc., and I think this only needs to be `Arc` rather than `Rc` because `mysql::Error` uses `Row`.)

**Comment by blackbeam** [comment]
Thanks!
Benchmarks shows that overhead is zero. Implementation already in my local master, but change are breaking so i will publish this together with fix for #43.
